File: csharp/CsSearch/CsSearch/Searcher.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace CsSearch
{
	public class Searcher
	{
		private readonly FileTypes _fileTypes;
		public SearchSettings Settings { get; private set; }
		public ConcurrentBag<SearchResult> Results { get; private set; }
		private Encoding TextFileEncoding { get; set; }

		private const int FileBatchSize = 255;

		public Searcher(SearchSettings settings)
		{
			Settings = settings;
			ValidateSettings();
			_fileTypes = new FileTypes();
			Results = new ConcurrentBag<SearchResult>();
		}

		private void ValidateSettings()
		{
			if (string.IsNullOrEmpty(Settings.StartPath))
				throw new SearchException("Startpath not defined");
			var expandedPath = FileUtil.ExpandPath(Settings.StartPath);
			if (FileUtil.IsDirectory(expandedPath))
			{
				if (!(new DirectoryInfo(expandedPath)).Exists)
					throw new SearchException("Startpath not found");
			}
			else
			{
				if (!(new FileInfo(expandedPath)).Exists)
					throw new SearchException("Startpath not found");
			}
			if (Settings.SearchPatterns.Count < 1)
				throw new SearchException("No search patterns defined");
			try
			{
				TextFileEncoding = Encoding.GetEncoding(Settings.TextFileEncoding);
			}
			catch (ArgumentException)
			{
				throw new SearchException("Invalid encoding");
			}
		}

		public bool IsSearchDirectory(DirectoryInfo d)
		{
			if (FileUtil.IsDotDir(d.Name))
				return true;
			if (Settings.ExcludeHidden && FileUtil.IsHidden(d))
				return false;
			if (Settings.InDirPatterns.Count > 0 &&
				!Settings.InDirPatterns.Any(p => p.Match(d.Name).Success))
				return false;
			if (Settings.OutDirPatterns.Count > 0 &&
				Settings.OutDirPatterns.Any(p => p.Match(d.Name).Success))
				return false;
			return true;
		}

		public bool IsSearchFile(FileInfo f)
		{
			if (Settings.InExtensions.Count > 0 &&
				!Settings.InExtensions.Contains(f.Extension))
				return false;
			if (Settings.OutExtensions.Count > 0 &&
				Settings.OutExtensions.Contains(f.Extension))
				return false;
			if (Settings.InFilePatterns.Count > 0 &&
				!Settings.InFilePatterns.Any(p => p.Match(f.Name).Success))
				return false;
			if (Settings.OutFilePatterns.Count > 0 &&
				Settings.OutFilePatterns.Any(p => p.Match(f.Name).Success))
				return false;
			return true;
		}

		public bool IsArchiveSearchFile(FileInfo f)
		{
			if (Settings.InArchiveExtensions.Count > 0 &&
				!Settings.InArchiveExtensions.Contains(f.Extension))
				return false;
			if (Settings.OutArchiveExtensions.Count > 0 &&
				Settings.OutArchiveExtensions.Contains(f.Extension))
				return false;
			if (Settings.InArchiveFilePatterns.Count > 0 &&
				!Settings.InArchiveFilePatterns.Any(p => p.Match(f.Name).Success))
				return false;
			if (Settings.OutArchiveFilePatterns.Count > 0 &&
				Settings.OutArchiveFilePatterns.Any(p => p.Match(f.Name).Success))
				return false;
			return true;
		}

		public IEnumerable<DirectoryInfo> GetSearchDirs(DirectoryInfo dir)
		{
			IEnumerable<DirectoryInfo> searchDirs = new List<DirectoryInfo>();
			try
			{
				searchDirs = dir.EnumerateDirectories().Where(IsSearchDirectory);
				return searchDirs.Aggregate(searchDirs, (current, d) => current.Concat(GetSearchDirs(d)));
			}
			catch (IOException e)
			{
				if (Settings.Verbose)
					Common.Log($"Error while accessing dir {FileUtil.ContractPath(dir.FullName)}: {e.Message}");
			}
			return searchDirs;
		}

		public bool FilterFile(FileInfo f)
		{
			if (FileUtil.IsHidden(f) && Settings.ExcludeHidden)
				return false;
			if (_fileTypes.IsArchiveFile(f))
			{
				return (Settings.SearchArchives && IsArchiveSearchFile(f));
			}
			return (!Settings.ArchivesOnly && IsSearchFile(f));
		}

		private SearchFile SearchFileFromFileInfo(FileInfo f)
		{
			return new SearchFile(new List<string>(), f.DirectoryName, f.Name, _fileTypes.GetFileType(f));
		}

		private IEnumerable<SearchFile> GetSearchFilesForDir(DirectoryInfo dir)
		{
			if (Settings.Debug)
			{
				Common.Log($"Getting search files under {FileUtil.ContractPath(dir.FullName)}");
			}
			IEnumerable<SearchFile> dirSearchFiles = new List<SearchFile>();
			try
			{
				dirSearchFiles = dir.EnumerateFiles().
					Where(FilterFile).
					Select((f,i) => SearchFileFromFileInfo(f));
			}
			catch (IOException e)
			{
				if (Settings.Verbose)
					Common.Log($"Error while accessing dir {FileUtil.ContractPath(dir.FullName)}: {e.Message}");
			}
			return dirSearchFiles;
		}

		public IEnumerable<SearchFile> GetSearchFiles(IEnumerable<DirectoryInfo> dirs)
		{
			var searchFiles = new List<SearchFile>();
			foreach (var d in dirs)
			{
				searchFiles.AddRange(GetSearchFilesForDir(d));
			}
			return searchFiles;
		}

		public void Search()
		{
			var expandedPath = FileUtil.ExpandPath(Settings.StartPath);
			if (FileUtil.IsDirectory(expandedPath))
			{
				var startDir = new DirectoryInfo(expandedPath);
				if (IsSearchDirectory(startDir))
				{
					SearchPath(startDir);
				}
				else
				{
					throw new SearchException("Startpath does not match search settings");
				}
			}
			else
			{
				var f = new FileInfo(expandedPath);
				if (FilterFile(f))
				{
					DoSearchFile(new SearchFile(f.DirectoryName, f.Name, _fileTypes.GetFileType(f)));
				}
				else
				{
					throw new SearchException("Startpath does not match search settings");
				}
			}
		}

		public void SearchPath(DirectoryInfo path)
		{
			var searchDirs = new List<DirectoryInfo> { path };
			if (Settings.Recursive)
			{
				searchDirs.AddRange(GetSearchDirs(path));
			}
			if (Settings.Verbose)
			{
				Common.Log($"Directories to be searched ({searchDirs.Count}):");
				foreach (var d in searchDirs)
				{
					Common.Log(FileUtil.ContractPath(d.FullName));
				}
				Common.Log("");
			}

			var searchFiles = GetSearchFiles(searchDirs).ToArray();
			if (Settings.Verbose)
			{
				Common.Log($"\nFiles to be searched ({searchFiles.Length}):");
				foreach (var f in searchFiles)
				{
					Common.Log(FileUtil.ContractPath(f.FullName));
				}
				Common.Log("");
			}

			var searched = 0;
			while (searchFiles.Length - searched > FileBatchSize)
			{
				SearchBatch(searchFiles.Skip(searched).Take(FileBatchSize).ToArray());
				searched += FileBatchSize;
			}
			if (searchFiles.Length > searched)
			{
				SearchBatch(searchFiles.Skip(searched).ToArray());
			}
		}

		private void SearchBatch(SearchFile[] searchFiles)
		{
			if (searchFiles.Length > 100)
			{
				SearchBatchConcurrent(searchFiles);
			}
			else
			{
				foreach (var f in searchFiles)
				{
					DoSearchFile(f);
				}
			}
		}

		private void SearchBatchConcurrent(SearchFile[] searchFiles)
		{
			var searchTasks = new Task[searchFiles.Length];
			for (var i = 0; i < searchFiles.Length; i++)
			{
				var searchFile = searchFiles[i];
				searchTasks[i] = Task.Factory.StartNew(() => DoSearchFile(searchFile));
			}
			Task.WaitAll(searchTasks);
		}

		public void DoSearchFile(SearchFile f)
		{
			switch (f.Type)
			{
				case FileType.Code:
				case FileType.Text:
				case FileType.Xml:
					SearchTextFile(f);
					break;
				case FileType.Binary:
					SearchBinaryFile(f);
					break;
				case FileType.Archive:
					Common.Log($"Skipping archive file {FileUtil.ContractPath(f.FullName)}");
					break;
				default:
				{
					if (Settings.Verbose)
					{
						Common.Log($"Skipping file {FileUtil.ContractPath(f.FullName)}");
					}

					break;
				}
			}
		}

		private void SearchTextFile(SearchFile f)
		{
			if (Settings.Verbose)
				Common.Log($"Searching text file {FileUtil.ContractPath(f.FullName)}");
			if (Settings.MultiLineSearch)
				SearchTextFileContents(f);
			else
				SearchTextFileLines(f);
		}

		private void SearchTextFileContents(SearchFile f)
		{
			try
			{
				var contents = FileUtil.GetFileContents(f, TextFileEncoding);
				var results = SearchContents(contents);
				foreach (var r in results)
				{
					r.File = f;
					AddSearchResult(r);
				}
			}
			catch (IOException e)
			{
				Common.Log(e.Message);
			}
		}

		private static IEnumerable<int> GetNewLineIndices(string text)
		{
			// to keep it simple, only get indices of '\n' and deal with '\r' later
			IList<int> newLineIndices = new List<int>();
			for (var i = 0; i < text.Length; i++)
			{
				if (text[i] == '\n')
					newLineIndices.Add(i);
			}
			return newLineIndices;
		}

		private static IEnumerable<int> GetStartLineIndices(IEnumerable<int> newLineIndices)
		{
			var startLineIndices = new List<int> {0};
			startLineIndices.AddRange(newLineIndices.Select(i => i + 1));
			return startLineIndices;
		}

		private static IEnumerable<int> GetEndLineIndices(string text,
			IEnumerable<int> newLineIndices)
		{
			var endLineIndices = new List<int>(newLineIndices) {text.Length - 1};
			return endLineIndices;
		}

		private IEnumerable<string> GetLinesBeforeFromContents(string content,
			IEnumerable<int> beforeStartIndices, IEnumerable<int> beforeEndIndices)
		{
			var linesBefore = new List<string>();
			if (Settings.LinesBefore > 0)
			{
				var starts = beforeStartIndices.Reverse().Take(Settings.LinesBefore).Reverse().ToList();
				var ends = beforeEndIndices.Reverse().Take(Settings.LinesBefore + 1).Reverse().Skip(1).ToList();
				linesBefore.AddRange(starts.Select((t, i) => content.Substring(t, ends[i] - t)));
			}
			return linesBefore;
		}

		private IEnumerable<string> GetLinesAfterFromContents(string content,
			IEnumerable<int> afterStartIndices, IEnumerable<int> afterEndIndices)
		{
			var linesAfter = new List<string>();
			if (Settings.LinesAfter > 0)
			{
				var starts = afterStartIndices.Take(Settings.LinesAfter).ToList();
				var ends = afterEndIndices.Skip(1).Take(Settings.LinesAfter).ToList();
				linesAfter.AddRange(starts.Select((t, i) => content.Substring(t, ends[i] - t)));
			}
			return linesAfter;
		}

		public IEnumerable<SearchResult> SearchContents(string contents)
		{
			var patternMatches = new Dictionary<Regex, int>();
			var results = new List<SearchResult>();
			var newLineIndices = GetNewLineIndices(contents);
			var startLineIndices = GetStartLineIndices(newLineIndices).ToList();
			var endLineIndices = GetEndLineIndices(contents, newLineIndices).ToList();

			foreach (var p in Settings.SearchPatterns)
			{
				var match = p.Match(contents);
				while (match.Success)
				{
					if (Settings.FirstMatch && patternMatches.ContainsKey(p))
					{
						break;
					}
					var matchIndex = match.Index;
					// get the start and end indices before the match index
					var beforeStartIndices = startLineIndices.TakeWhile(i => i <= matchIndex).ToList();
					var beforeEndIndices = endLineIndices.TakeWhile(i => i < matchIndex).ToList();
					// add another end line index if it exists or the last index of the string
					if (endLineIndices.Count > beforeEndIndices.Count)
						beforeEndIndices.ToList().Add(endLineIndices[beforeEndIndices.Count]);
					else
						beforeEndIndices.ToList().Add(contents.Length - 1);
					var afterStartIndices = startLineIndices.SkipWhile(i => i <= matchIndex).ToList();
					var afterEndIndices = endLineIndices.SkipWhile(i => i <= matchIndex).ToList();
					var lineNum = beforeStartIndices.ToList().Count;
					var startLineIndex = beforeStartIndices.Max();
					var endLineIndex = afterStartIndices.Min() - 1;
					var line = contents.Substring(startLineIndex, endLineIndex - startLineIndex);
					var linesBefore = GetLinesBeforeFromContents(contents,
						beforeStartIndices.AsEnumerable().Reverse().Skip(1).Reverse(),
						beforeEndIndices);
					var linesAfter = GetLinesAfterFromContents(contents, afterStartIndices, afterEndIndices);
					if ((linesBefore.ToList().Count == 0 || LinesBeforeMatch(linesBefore))
						&&
						(linesAfter.ToList().Count == 0 || LinesAfterMatch(linesAfter)))
						results.Add(new SearchResult(
						p,
						null,
						lineNum,
						match.Index - startLineIndex + 1,
						match.Index - startLineIndex + match.Length + 1,
						line,
						new List<string>(linesBefore),
						new List<string>(linesAfter)));
					patternMatches[p] = 1;

					match = match.NextMatch();
				}
			}
			return results;
		}

		private void SearchTextFileLines(SearchFile f)
		{
			try
			{
				var enumerableLines = FileUtil.EnumerableStringFromFile(f, TextFileEncoding);
				var results = SearchLines(enumerableLines);

				foreach (var r in results)
				{
					r.File = f;
					AddSearchResult(r);
				}
			}
			catch (IOException e)
			{
				Common.Log(e.Message);
			}
		}

		private static bool AnyMatchesAnyPattern(IEnumerable<string> strings,
			IEnumerable<Regex> patterns)
		{
			return strings.Any(s => MatchesAnyPattern(s, patterns));
		}

		private static bool MatchesAnyPattern(string s, IEnumerable<Regex> patterns)
		{
			return !string.IsNullOrEmpty(s) && patterns.Any(p => p.Match(s).Success);
		}

		private static bool LinesMatch(IEnumerable<string> lines,
			ICollection<Regex> inPatterns, ICollection<Regex> outPatterns)
		{
			return ((inPatterns.Count == 0 || AnyMatchesAnyPattern(lines, inPatterns))
				&& (outPatterns.Count == 0 || !AnyMatchesAnyPattern(lines, outPatterns)));
		}

		private bool LinesBeforeMatch(IEnumerable<string> linesBefore)
		{
			return LinesMatch(linesBefore, Settings.InLinesBeforePatterns,
				Settings.OutLinesBeforePatterns);
		}

		private bool LinesAfterMatch(IEnumerable<string> linesAfter)
		{
			return LinesMatch(linesAfter, Settings.InLinesAfterPatterns,
				Settings.OutLinesAfterPatterns);
		}

		public IEnumerable<SearchResult> SearchLines(IEnumerable<string> lines)
		{
			var patternMatches = new Dictionary<Regex, int>();
			var results = new List<SearchResult>();
			var lineNum = 0;
			var linesBefore = new Queue<string>();
			var linesAfter = new Queue<string>();
			var lineEnumerator = lines.GetEnumerator();
			var stop = false;

			while ((lineEnumerator.MoveNext() || linesAfter.Count > 0) && !stop)
			{
				lineNum++;
				var line = linesAfter.Count > 0 ? linesAfter.Dequeue() : lineEnumerator.Current;
				if (Settings.LinesAfter > 0)
				{
					while (linesAfter.Count < Settings.LinesAfter && lineEnumerator.MoveNext())
					{
						linesAfter.Enqueue(lineEnumerator.Current);
					}
				}

				if ((Settings.LinesBefore == 0 || linesBefore.Count == 0 || LinesBeforeMatch(linesBefore))
					&&
					(Settings.LinesAfter == 0 || linesAfter.Count == 0 || LinesAfterMatch(linesAfter)))
				{
					foreach (var p in Settings.SearchPatterns)
					{
						var matches = p.Matches(line);
						foreach (Match match in matches)
						{
							if (Settings.FirstMatch && patternMatches.ContainsKey(p))
							{
								stop = true;
								break;
							}
							results.Add(new SearchResult(p,
								null,
								lineNum,
								match.Index + 1,
								match.Index + match.Length + 1,
								line,
								new List<string>(linesBefore),
								new List<string>(linesAfter)));
							patternMatches[p] = 1;
						}
					}
				}
				if (Settings.LinesBefore > 0)
				{
					if (linesBefore.Count == Settings.LinesBefore)
					{
						linesBefore.Dequeue();
					}
					if (linesBefore.Count < Settings.LinesBefore)
					{
						linesBefore.Enqueue(line);
					}
				}
			}
			return results;
		}

		private void SearchBinaryFile(SearchFile f)
		{
			if (Settings.Verbose)
				Common.Log($"Searching binary file {FileUtil.ContractPath(f.FullName)}");
			try
			{
				// get the binary bytes in a single-byte encoding (will break if UTF8)
				//using (var sr = new StreamReader(f.FullName, Encoding.GetEncoding("ISO_8859_1")))
				using (var sr = new StreamReader(f.FullName))
				{
					var contents = sr.ReadToEnd();
					foreach (var p in Settings.SearchPatterns)
					{
						var matches = p.Matches(contents).Cast<Match>();
						if (Settings.FirstMatch)
						{
							matches = matches.Take(1);
						}
						foreach (var m in matches)
						{
							AddSearchResult(new SearchResult(
								p,
								f,
								0,
								m.Index + 1,
								m.Index + m.Length + 1,
								null));
						}
					}
				}
			}
			catch (IOException e)
			{
				Common.Log(e.Message);
			}
		}

		private void AddSearchResult(SearchResult searchResult)
		{
			Results.Add(searchResult);
		}

		private IList<SearchResult> GetSortedSearchResults()
		{
			IComparer<SearchResult> comparer = new SearchResultsComparer();
			var sorted = Results.ToList();
			sorted.Sort(comparer);
			return sorted;
		}

		public void PrintResults()
		{
			var sortedResults = GetSortedSearchResults();
			Common.Log($"Search results ({sortedResults.Count}):");
			foreach (var searchResult in sortedResults)
			{
				Common.Log(searchResult.ToString(Settings));
			}
		}

		public IEnumerable<DirectoryInfo> GetMatchingDirs()
		{
			return new List<DirectoryInfo>(
				Results.Select(r => r.File.FilePath).
				Distinct().Select(d => new DirectoryInfo(d)).
				OrderBy(d => d.FullName));
		}

		public void PrintMatchingDirs()
		{
			var matchingDirs = GetMatchingDirs();
			Common.Log($"\nDirectories with matches ({matchingDirs.Count()}):");
			foreach (var d in matchingDirs)
			{
				Common.Log(FileUtil.GetRelativePath(d.FullName, Settings.StartPath));
			}
		}

		public IEnumerable<FileInfo> GetMatchingFiles()
		{
			return new List<FileInfo>(
				Results.Select(r => r.File.PathAndName).
				Distinct().Select(f => new FileInfo(f)).
				OrderBy(d => d.FullName));
		}

		public void PrintMatchingFiles()
		{
			var matchingFiles = GetMatchingFiles();
			Common.Log($"\nFiles with matches ({matchingFiles.Count()}):");
			foreach (var f in matchingFiles)
			{
				Common.Log(FileUtil.GetRelativePath(f.FullName, Settings.StartPath));
			}
		}

		public IEnumerable<string> GetMatchingLines()
		{
			var lines = Results.Select(r => r.Line.Trim()).ToList();
			if (Settings.UniqueLines)
			{
				lines = new HashSet<string>(lines).ToList();
			}
			lines.Sort(new CaseInsensitiveComparer());
			return lines;
		}

		public void PrintMatchingLines()
		{
			var matchingLines = GetMatchingLines();
			var hdrText = Settings.UniqueLines ? "Unique lines with matches" : "Lines with matches";
			Common.Log($"\n{hdrText} ({matchingLines.Count()}):");
			foreach (var m in matchingLines)
			{
				Common.Log(m);
			}
		}
	}

	internal class CaseInsensitiveComparer : IComparer<string>
	{
		public int Compare(string a, string b)
		{
			return a.ToUpper().CompareTo(b.ToUpper());
		}
	}
}

File: fsharp/FsSearch/Searcher.fs
﻿namespace FsSearch

open System
open System.Collections.Generic
open System.Diagnostics
open System.IO
open System.Text.RegularExpressions

type Searcher (settings : SearchSettings) =
    let _settings = settings
    let _fileTypes = new FileTypes()
    let _results = new List<SearchResult>()
    let _fileSet = new HashSet<FileInfo>()

    // read-only member properties
    member this.Settings = _settings
    member this.Results = _results
    member this.FileSet = _fileSet

    // member methods
    member this.ValidateSettings () : string list =
        let errs = new List<string>()
        if String.IsNullOrEmpty _settings.StartPath then
            errs.Add("Startpath not defined")
        else if not (FileUtil.IsDirectory _settings.StartPath) && not (new FileInfo(_settings.StartPath)).Exists then
            errs.Add("Startpath not found")
        if _settings.SearchPatterns.Count = 0 then errs.Add("No search patterns defined")
        errs |> List.ofSeq

    member this.MatchesAnyPattern (s : string) (patterns : ISet<Regex>) =
        Seq.exists (fun p -> (p:Regex).Match(s).Success) patterns

    member this.AnyMatchesAnyPattern (slist : string seq) (patterns : ISet<Regex>) =
        Seq.exists (fun s -> this.MatchesAnyPattern s patterns) slist

    member this.IsSearchDir (d : DirectoryInfo) =
        let elems = d.FullName.Split('/', '\\') |> Seq.filter (fun s -> not (String.IsNullOrEmpty s))
        let hiddenOk = not _settings.ExcludeHidden || not (Seq.exists (fun e -> FileUtil.IsHidden e) elems)
        let matchesIn = (Seq.isEmpty _settings.InDirPatterns || this.AnyMatchesAnyPattern elems _settings.InDirPatterns)
        let notMatchesOut = (Seq.isEmpty _settings.OutDirPatterns || not (this.AnyMatchesAnyPattern elems _settings.OutDirPatterns))
        hiddenOk && matchesIn && notMatchesOut

    member this.IsSearchFile (f : FileInfo) =
        (Seq.isEmpty _settings.InExtensions ||
         Seq.exists (fun x -> x = f.Extension) _settings.InExtensions) &&
        (Seq.isEmpty _settings.OutExtensions ||
         not (Seq.exists (fun x -> x = f.Extension) _settings.OutExtensions)) &&
        (Seq.isEmpty _settings.InFilePatterns ||
         Seq.exists (fun p -> (p:Regex).Match(f.Name).Success) _settings.InFilePatterns) &&
        (Seq.isEmpty _settings.OutFilePatterns ||
         not (Seq.exists (fun p -> (p:Regex).Match(f.Name).Success) _settings.OutFilePatterns))

    member this.IsArchiveSearchFile (f : FileInfo) =
        (Seq.isEmpty _settings.InArchiveExtensions ||
         Seq.exists (fun x -> x = f.Extension) _settings.InArchiveExtensions) &&
        (Seq.isEmpty _settings.OutArchiveExtensions ||
         not (Seq.exists (fun x -> x = f.Extension) _settings.OutArchiveExtensions)) &&
        (Seq.isEmpty _settings.InArchiveFilePatterns ||
         Seq.exists (fun p -> (p:Regex).Match(f.Name).Success) _settings.InArchiveFilePatterns) &&
        (Seq.isEmpty _settings.OutArchiveFilePatterns ||
         not (Seq.exists (fun p -> (p:Regex).Match(f.Name).Success) _settings.OutArchiveFilePatterns))

    member this.FilterFile (f: FileInfo) = 
        if FileUtil.IsHidden f.Name && _settings.ExcludeHidden then
            false
        else if _fileTypes.IsArchiveFile f then
            _settings.SearchArchives && this.IsArchiveSearchFile f
        else
            not _settings.ArchivesOnly && this.IsSearchFile f

    member this.GetSearchDirs (dir : DirectoryInfo) : DirectoryInfo list =
        try
            let dirs = 
                dir.EnumerateDirectories()
                |> Seq.filter (fun d -> this.IsSearchDir (d:DirectoryInfo))
                |> List.ofSeq
            dirs @ List.fold (fun ds d -> (ds @ (this.GetSearchDirs d))) [] dirs
        with
        | :? IOException as e ->
            Common.Log (sprintf "Error while accessing dir %s: %s" dir.FullName e.Message)
            List.empty<DirectoryInfo>

    member this.GetSearchFilesForDir (dir : DirectoryInfo) =
        try
            dir.EnumerateFiles()
            |> Seq.filter (fun f -> this.FilterFile (f:FileInfo))
            |> List.ofSeq
        with
        | :? IOException as e ->
            Common.Log (sprintf "Error while accessing dir %s: %s" dir.FullName e.Message)
            List.empty<FileInfo>

    member this.GetSearchFiles (dirs : DirectoryInfo list) =
        let rec getSearchFiles (files : FileInfo list, dirs : DirectoryInfo list) =
            match dirs with
            | [] -> files
            | d :: ds -> getSearchFiles(files @ (this.GetSearchFilesForDir d), ds)
        getSearchFiles([], dirs)

    member this.Search () =
        if FileUtil.IsDirectory(_settings.StartPath) then
            let startDir = new DirectoryInfo(_settings.StartPath)
            if this.IsSearchDir startDir then
                this.SearchPath startDir
            else
                raise <| Exception ("Startpath does not match search settings")
        else
            this.SearchFile (new FileInfo(_settings.StartPath))

    member this.SearchPath (startDir : DirectoryInfo) =
        let dirs =
            if _settings.Recursive then
                startDir :: this.GetSearchDirs(startDir)
            else
                [startDir]
        if _settings.Verbose then
            printfn "\nDirectories to be searched (%d):" dirs.Length
            List.iter (fun d -> Common.Log (sprintf "%s " (d:DirectoryInfo).FullName)) dirs
        
        let files = this.GetSearchFiles(dirs)
        if _settings.Verbose then
            Common.Log (sprintf "\nFiles to be searched (%d):" files.Length)
            Seq.iter (fun f -> Common.Log (sprintf "%s " (f:FileInfo).FullName)) files
            printfn ""

        for f in files do
            this.SearchFile f

    member this.SearchFile (f : FileInfo) =
        match _fileTypes.GetFileType f with
        | FileType.Archive -> Common.Log (sprintf "Archive file searching not currently supported")
        | FileType.Binary -> this.SearchBinaryFile f
        | FileType.Text -> this.SearchTextFile f
        | FileType.Unknown -> Common.Log (sprintf "Skipping file of unknown type")
        | _ -> Common.Log (sprintf "Skipping file of indeterminate type (this shouldn't happen): %s" f.FullName)

    member this.SearchTextFile (f : FileInfo) =
        if _settings.Verbose then
            Common.Log (sprintf "Searching text file %s" f.FullName)
        if _settings.MultiLineSearch then
            this.SearchTextFileContents f
        else
            this.SearchTextFileLines f

    member this.SearchTextFileContents (f : FileInfo) =
        try
            let contents = FileUtil.GetFileContents f.FullName
            let results = this.SearchContents(contents)
            for r:SearchResult in results do
                r.File <- f
                this.AddSearchResult r
        with
        | :? IOException as e -> printfn "%s" e.Message

    member this.GetLineIndices (s : string) =
        let newLineIndices =
            Array.zip [|0 .. s.Length - 1|] (s.ToCharArray())
            |> Array.filter (fun x -> snd x = '\n')
            |> Array.map (fun x -> fst x)
            |> List.ofArray
        let startLineIndices =
             0 :: List.map (fun i -> i + 1) newLineIndices
             |> Seq.takeWhile (fun i -> i < s.Length)
             |> List.ofSeq
        let endLineIndices = newLineIndices
        List.zip startLineIndices endLineIndices

    member this.LinesMatch (lines : string seq, inPatterns : ISet<Regex>,
                            outPatterns : ISet<Regex>) =
        (inPatterns.Count = 0 || (this.AnyMatchesAnyPattern lines inPatterns))
        &&
        (outPatterns.Count = 0  || (not (this.AnyMatchesAnyPattern lines outPatterns)))

    member this.LinesBeforeMatch (linesBefore : string seq) =
        this.LinesMatch (linesBefore, settings.InLinesBeforePatterns,
                         settings.OutLinesBeforePatterns)

    member this.LinesAfterMatch (linesAfter : string seq) =
        this.LinesMatch(linesAfter, settings.InLinesAfterPatterns,
                        settings.OutLinesAfterPatterns)

    member this.TakeRight (ss : 'T list) (num : int) =
        if ss.Length < num then
            ss
        else
            ss
            |> List.ofSeq
            |> List.rev
            |> Seq.take num
            |> List.ofSeq
            |> List.rev

    member this.SearchContents (s : string) =
        let patternMatches = new Dictionary<Regex, int>()
        let results = new List<SearchResult>()
        let lineIndices = this.GetLineIndices s
        for p in _settings.SearchPatterns do
            let mutable m = p.Match s
            let mutable stop = false
            while m.Success && not stop do
                let matchStartIndex = m.Index
                let matchEndIndex = m.Index + m.Length
                let startLineIndex = Seq.filter (fun i -> fst i < matchStartIndex) lineIndices |> Seq.max |> fst
                //Common.Log (sprintf "startLineIndex: %d" startLineIndex)
                let endLineIndex = Seq.filter (fun i -> snd i > matchStartIndex) lineIndices |> Seq.min |> snd
                //Common.Log (sprintf "endLineIndex: %d" endLineIndex)
                let beforeIndices = Seq.takeWhile (fun i -> fst i < startLineIndex) lineIndices |> List.ofSeq
                //Common.Log (Common.ListToString("beforeIndices", beforeIndices))
                let afterIndices = Seq.skipWhile (fun i -> fst i <= matchStartIndex) lineIndices |> List.ofSeq
                let lineNum = beforeIndices.Length + 1
                let line = s.Substring(startLineIndex, endLineIndex - startLineIndex)
                let beforeLineIndices = this.TakeRight beforeIndices settings.LinesBefore
                //Common.Log (Common.ListToString("beforeLineIndices", beforeLineIndices))
                let linesBefore =
                    if settings.LinesBefore > 0 then
                        beforeLineIndices
                        |> List.map (fun (x,y) -> s.Substring(x, y - x))
                    else
                        []
                let linesAfter =
                    if settings.LinesAfter > 0 then
                        if afterIndices.Length < settings.LinesAfter then
                            afterIndices
                            |> List.ofSeq
                            |> List.map (fun (x,y) -> s.Substring(x, y - x))
                        else
                            afterIndices
                            |> Seq.take(settings.LinesAfter)
                            |> List.ofSeq
                            |> List.map (fun (x,y) -> s.Substring(x, y - x))
                    else
                        []
                if (linesBefore.Length = 0 || this.LinesBeforeMatch linesBefore) &&
                   (linesAfter.Length = 0 || this.LinesAfterMatch linesAfter) then
                    results.Add(new SearchResult(p,
                                                 null,
                                                 lineNum,
                                                 matchStartIndex - startLineIndex + 1,
                                                 matchEndIndex - startLineIndex + 1,
                                                 line,
                                                 linesBefore,
                                                 linesAfter))
                    if _settings.FirstMatch then
                        stop <- true
                m <- m.NextMatch()
        results

    member this.SearchTextFileLines (f : FileInfo) =
        try
            let lines = FileUtil.GetFileLines f.FullName
            let results = this.SearchLines lines
            for r:SearchResult in results do
                r.File <- f
                this.AddSearchResult r
        with
        | :? IOException as e -> printfn "%s" e.Message

    member this.SearchLines (lines : string seq) =
        let patternMatches = new Dictionary<Regex, int>()
        let results = new List<SearchResult>()
        let mutable lineNum = 0
        for line in lines do
            lineNum <- lineNum + 1
            for p in _settings.SearchPatterns do
                if not _settings.FirstMatch || not (patternMatches.ContainsKey p) then
                    let matches = p.Matches(line)
                    if matches.Count > 0 then
                        if not (patternMatches.ContainsKey p) then
                            patternMatches.Add(p, 1)
                        if _settings.FirstMatch then
                            let m = matches.[0]
                            let matchStartIndex = m.Index + 1
                            let matchEndIndex = m.Index + m.Length + 1
                            results.Add(new SearchResult(p, null, lineNum, matchStartIndex, matchEndIndex, line, [], []))
                        else
                            for m in matches do
                                let matchStartIndex = m.Index + 1
                                let matchEndIndex = m.Index + m.Length + 1
                                results.Add(new SearchResult(p, null, lineNum, matchStartIndex, matchEndIndex, line, [], []))
        results

    member this.SearchBinaryFile (f : FileInfo) =
        if _settings.Verbose then
            Common.Log (sprintf "Searching binary file %s" f.FullName)
        try
            use sr = new StreamReader (f.FullName)
            let contents = sr.ReadToEnd()
            for p in Seq.filter (fun p -> (p:Regex).Match(contents).Success) _settings.SearchPatterns do
                let mutable m = p.Match contents
                let mutable stop = false
                while m.Success && not stop do
                    let matchStartIndex = m.Index + 1
                    let matchEndIndex = m.Index + m.Length + 1
                    this.AddSearchResult (new SearchResult(p,
                                                           f,
                                                           0,
                                                           matchStartIndex,
                                                           matchEndIndex,
                                                           null,
                                                           [],
                                                           []))
                    if _settings.FirstMatch then
                        stop <- true
                    m <- m.NextMatch()
        with
        | :? IOException as ex -> printfn "%s" (ex.Message)

    member this.AddSearchResult (searchResult : SearchResult) =
        let fileadd = _fileSet.Add(searchResult.File)
        _results.Add(searchResult)

    member this.GetSortedResults : SearchResult list = 
        this.Results
        |> Seq.sortBy (fun r -> r.SortKey)
        |> List.ofSeq

    member this.PrintResults = 
        let sortedResults = this.GetSortedResults
        Common.Log (sprintf "\nSearch results (%d):" this.Results.Count)
        sortedResults
        |> Seq.iter (fun r -> Common.Log (sprintf "%s" (r.ToString())))

    member this.GetMatchingDirs : DirectoryInfo list = 
        this.Results
        |> Seq.map (fun r -> r.File.Directory)
        |> Seq.distinctBy (fun d -> d.FullName)
        |> Seq.sortBy (fun d -> d.FullName)
        |> List.ofSeq

    member this.PrintMatchingDirs = 
        let dirs = this.GetMatchingDirs
        Common.Log (sprintf "\nDirectories with matches (%d):" dirs.Length)
        for d in dirs do
            printfn "%s" d.FullName

    member this.GetMatchingFiles : FileInfo list = 
        this.Results
        |> Seq.map (fun r -> r.File)
        |> Seq.distinctBy (fun f -> f.FullName)
        |> Seq.sortBy (fun f -> f.FullName)
        |> List.ofSeq

    member this.PrintMatchingFiles = 
        let files = this.GetMatchingFiles
        Common.Log (sprintf "\nFiles with matches (%d):" files.Length)
        for f in files do
            printfn "%s" f.FullName

    member this.GetMatchingLines : string list = 
        let lines =
            this.Results
            |> Seq.map (fun r -> r.Line.Trim())
        if _settings.UniqueLines then
            Seq.sortBy (fun (s : string) -> s.ToUpper()) lines
            |> Seq.distinct
            |> List.ofSeq
        else
            Seq.sortBy (fun (s : string) -> s.ToUpper()) lines
            |> List.ofSeq

    member this.PrintMatchingLines = 
        let lines = this.GetMatchingLines
        let title =
            if _settings.UniqueLines then "Unique lines with matches"
            else "Lines with matches"
        Common.Log (sprintf "\n%s (%d):" title lines.Length)
        for l in lines do
            printfn "%s" l


    ;;
